### PR TITLE
feat: ngql websocket

### PR DESCRIPTION
**Pull request by huaxiabuluo**
## What type of PR is this?
- [ ] bug
- [x] feature
- [ ] enhancement

### CODE REVIEW

File: app/stores/global.ts [hunk 1]
```diff
@@ -3,8 +3,10 @@ import cookies from 'js-cookie';
 import { message } from 'antd';
 import { Base64 } from 'js-base64';
 import { getI18n } from '@vesoft-inc/i18n';
-import service from '@app/config/service';
+import intl from 'react-intl-universal';
```
Comment (hetao92) [inline app/stores/global.ts]: use our i18n

File: server/api/studio/studio.go [hunk 2]
```diff
@@ -70,6 +71,22 @@ func main() {
...
+	go hub.Run()
+	server.AddRoute(rest.Route{
+		Method: http.MethodGet,
+		Path:   "/nebula_ws",
+		Handler: func(w http.ResponseWriter, r *http.Request) {
+			clientInfo := &auth.AuthData{}
+			tokenCookie, err := r.Cookie(svcCtx.Config.Auth.TokenName)
+			if err == nil {
+				clientInfo, _ = auth.Decode(tokenCookie.Value, svcCtx.Config.Auth.AccessSecret)
+			}
```
Comment (kqzh) [inline server/api/studio/studio.go:85]: if err != nil, need handle err? and why ingore auth.Decode err

File: server/api/studio/pkg/ws/client.go [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+
+	// Buffered channel of outbound messages.
+	send chan []byte
+}
+
+func (c *Client) runNgql(msgReceived *MessageReceive) {
+	msgPost := MessagePost{}
+	msgPost.Header.MsgId = msgReceived.Header.MsgId
+	msgPost.Header.SendTime = time.Now().UnixMilli()
+	msgPost.Body.MsgType = msgReceived.Body.MsgType
```
Comment (kqzh) [inline server/api/studio/pkg/ws/client.go]: merge into one declaration? like
```go
msgPost := MessagePost{
    ...
}
```

File: server/api/studio/pkg/ws/client.go [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+		gql, _ = reqGql.(string)
+	}
+
+	reqParamList := msgReceived.Body.Content["paramList"]
+	if reqParamList != nil && reflect.TypeOf(reqParamList).Kind() == reflect.Slice {
+		s := reqParamList.([]interface{})
+		for i := 0; i < len(s); i++ {
+			paramList = append(paramList, s[i].(string))
+		}
+	}
```
Comment (kqzh) [inline server/api/studio/pkg/ws/client.go]: use
```go
reqParamList, ok := msgReceived.Body.Content["paramList"].([]interface{})
if ok {
    paramList = append(paramList, reqParamList[i].(string)...)
}

```
instead ?

File: server/api/studio/pkg/ws/client.go [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+	msgPost.Header.MsgId = msgReceived.Header.MsgId
+	msgPost.Header.SendTime = time.Now().UnixMilli()
+	msgPost.Body.MsgType = msgReceived.Body.MsgType
+
+	gql, paramList := "", []string{}
+
+	reqGql := msgReceived.Body.Content["gql"]
+	if reqGql != nil {
+		gql, _ = reqGql.(string)
+	}
```
Comment (kqzh) [inline server/api/studio/pkg/ws/client.go]: use
```go
reqGql, ok := msgReceived.Body.Content["gql"].(string)
if ok {
  gql = reqGql
}
```
instead ?

File: server/api/studio/pkg/ws/client.go [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+	reqParamList := msgReceived.Body.Content["paramList"]
+	if reqParamList != nil && reflect.TypeOf(reqParamList).Kind() == reflect.Slice {
+		s := reqParamList.([]interface{})
+		for i := 0; i < len(s); i++ {
+			paramList = append(paramList, s[i].(string))
+		}
+	}
+
+	execute, _, err := dao.Execute(c.clientInfo.NSID, gql, paramList)
+
```
Comment (kqzh) [inline server/api/studio/pkg/ws/client.go]: delete this empty line?

File: server/api/studio/pkg/ws/client.go [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+	if err != nil {
+		content := map[string]any{
+			"code":    base.Error,
+			"message": err.Error(),
+		}
+		if auth.IsSessionError(err) {
+			content["code"] = ecode.ErrSession.GetCode()
+		}
+		msgPost.Body.Content = &content
```
Comment (kqzh) [inline server/api/studio/pkg/ws/client.go:103]: why add &?

File: server/api/studio/pkg/ws/client.go [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+
+	msgSend, _ := json.Marshal(msgPost)
+	c.send <- msgSend
+}
+
+func (c *Client) runBatchNgql(msgReceived *MessageReceive) {
+	msgPost := MessagePost{}
+	msgPost.Header.MsgId = msgReceived.Header.MsgId
+	msgPost.Header.SendTime = time.Now().UnixMilli()
+	msgPost.Body.MsgType = msgReceived.Body.MsgType
```
Comment (kqzh) [inline server/api/studio/pkg/ws/client.go]: same as runNgql

File: server/api/studio/pkg/ws/client.go [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+		}
+
+		resContentData = append(resContentData, gqlRes)
+	}
+
+	msgPost.Body.Content = map[string]any{
+		"code":    base.Success,
+		"data":    resContentData,
+		"message": "Success",
+	}
```
Comment (kqzh) [inline server/api/studio/pkg/ws/client.go:179]: execute batch ngql will never failed?

File: server/api/studio/pkg/ws/hub.go [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+		case message := <-h.broadcast:
+			for client := range h.clients {
+				select {
+				case client.send <- message:
+				default:
+					close(client.send)
+					delete(h.clients, client)
+				}
+			}
+		}
```
Comment (kqzh) [inline server/api/studio/pkg/ws/hub.go]: unused now?

File: server/api/studio/pkg/ws/client.go [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+	msgPost := MessagePost{}
+	msgPost.Header.MsgId = msgReceived.Header.MsgId
+	msgPost.Header.SendTime = time.Now().UnixMilli()
+	msgPost.Body.MsgType = msgReceived.Body.MsgType
+
+	gql, paramList := "", []string{}
+
+	reqGql := msgReceived.Body.Content["gql"]
+	if reqGql != nil {
+		gql, _ = reqGql.(string)
```
Comment (veezhang) [inline server/api/studio/pkg/ws/client.go]: I think that it's should not to be ignored. How about you ?

File: server/api/studio/pkg/ws/client.go [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+import (
+	"bytes"
+	"encoding/json"
+	"log"
+	"net/http"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/gorilla/websocket"
```
Comment (veezhang) [inline server/api/studio/pkg/ws/client.go:11]: Forgot to lint ?

File: server/api/studio/pkg/ws/client.go [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+	writeWait = 10 * time.Second
+
+	// Time allowed to read the next pong message from the peer.
+	pongWait = 60 * time.Second
+
+	// Send pings to peer with this period. Must be less than pongWait.
+	pingPeriod = (pongWait * 9) / 10
+
+	// Maximum message size allowed from peer.
+	maxMessageSize = 16 * 1024
```
Comment (veezhang) [inline server/api/studio/pkg/ws/client.go]: Is `16 * 1024` enough?

File: server/api/studio/pkg/ws/client.go [hunk 14]
```diff
@@ -0,0 +1,285 @@
...
+	reqGql := msgReceived.Body.Content["gql"]
+	if reqGql != nil {
+		gql, _ = reqGql.(string)
+	}
+
+	reqParamList := msgReceived.Body.Content["paramList"]
+	if reqParamList != nil && reflect.TypeOf(reqParamList).Kind() == reflect.Slice {
+		s := reqParamList.([]interface{})
+		for i := 0; i < len(s); i++ {
+			paramList = append(paramList, s[i].(string))
```
Comment (veezhang) [inline server/api/studio/pkg/ws/client.go]: `s[i]` maybe not a `string` ?

File: server/api/studio/pkg/ws/client.go (hunk 4 above)
Comment (veezhang) [reply server/api/studio/pkg/ws/client.go]: @kqzh `[]any`?

File: server/api/studio/pkg/ws/client.go [hunk 15]
```diff
@@ -0,0 +1,285 @@
...
+		msgPost.Body.Content = &content
+	} else {
+		msgPost.Body.Content = map[string]any{
+			"code":    base.Success,
+			"data":    &execute,
+			"message": "Success",
+		}
+	}
+
+	msgSend, _ := json.Marshal(msgPost)
```
Comment (veezhang) [inline server/api/studio/pkg/ws/client.go:112]: Why not check the error?

File: server/api/studio/pkg/ws/client.go [hunk 16]
```diff
@@ -0,0 +1,285 @@
...
+	gqls, paramList := []string{}, []string{}
+	resContentData := make([]map[string]interface{}, 0)
+
+	reqGqls := msgReceived.Body.Content["gqls"]
+	if reqGqls != nil && reflect.TypeOf(reqGqls).Kind() == reflect.Slice {
+		s := reqGqls.([]interface{})
+		for i := 0; i < len(s); i++ {
+			gqls = append(gqls, s[i].(string))
+		}
+	}
```
Comment (veezhang) [inline server/api/studio/pkg/ws/client.go]: Maybe add some functions for `MessageReceiveBody` is more clearer?
And check the type in those functions.
```
func (b *MessageReceiveBody) GetGqls() (string, error) {
}
func (b *MessageReceiveBody) GetParamList() ([]string, error) {
}
// and so on 
```

File: server/api/studio/pkg/ws/client.go [hunk 17]
```diff
@@ -0,0 +1,285 @@
...
+	for {
+		_, message, err := c.conn.ReadMessage()
+		if err != nil {
+			if websocket.IsUnexpectedCloseError(err, websocket.CloseGoingAway, websocket.CloseAbnormalClosure) {
+				log.Printf("error: %v", err)
+			}
+			break
+		}
+
+		msgReceivedStr := bytes.TrimSpace(bytes.Replace(message, newline, space, -1))
```
Comment (veezhang) [inline server/api/studio/pkg/ws/client.go]: Why need `bytes.Replace(message, newline, space, -1)`?

File: server/api/studio/pkg/ws/client.go [hunk 18]
```diff
@@ -0,0 +1,285 @@
...
+				return
+			}
+
+			w, err := c.conn.NextWriter(websocket.TextMessage)
+			if err != nil {
+				return
+			}
+			w.Write(message)
+
+			// Add queued chat messages to the current websocket message.
```
Comment (veezhang) [inline server/api/studio/pkg/ws/client.go:254]: chat messages?

File: server/api/studio/pkg/ws/client.go [hunk 19]
```diff
@@ -0,0 +1,285 @@
...
+			n := len(c.send)
+			for i := 0; i < n; i++ {
+				w.Write(newline)
```
Comment (veezhang) [inline server/api/studio/pkg/ws/client.go:257]: Why need to write `newline` ?

File: server/api/studio/pkg/ws/client.go [hunk 20]
```diff
@@ -0,0 +1,285 @@
...
+		case <-ticker.C:
+			c.conn.SetWriteDeadline(time.Now().Add(writeWait))
+			if err := c.conn.WriteMessage(websocket.PingMessage, nil); err != nil {
+				return
+			}
+		}
+	}
+}
+
+// ServeWs handles websocket requests from the peer.
```
Comment (veezhang) [inline server/api/studio/pkg/ws/client.go:273]: `ServeWs` => `ServeWebSocket`?

File: server/api/studio/pkg/ws/hub.go [hunk 21]
```diff
@@ -0,0 +1,49 @@
+package ws
+
+// Hub maintains the set of active clients and broadcasts messages to the
+// clients.
+type Hub struct {
+	// Registered clients.
+	clients map[*Client]bool
+
+	// Inbound messages from the clients.
+	broadcast chan []byte
```
Comment (veezhang) [inline server/api/studio/pkg/ws/hub.go]: For what?

File: server/api/studio/pkg/ws/hub.go (hunk 10 above)
Comment (huaxiabuluo) [reply server/api/studio/pkg/ws/hub.go]: has been commented out, may be used in future

File: server/api/studio/pkg/ws/client.go (hunk 11 above)
Comment (huaxiabuluo) [reply server/api/studio/pkg/ws/client.go]: `msgReceived.Body.Content["gql"]` may be a `nil` value in some case, only `paramList` is passed in request data

File: server/api/studio/pkg/ws/client.go (hunk 9 above)
Comment (huaxiabuluo) [reply server/api/studio/pkg/ws/client.go:179]: yeah, single ngql execution may be error res, batch req should never fail unless network anomaly or other unexpected server error

File: server/api/studio/studio.go (hunk 2 above)
Comment (huaxiabuluo) [reply server/api/studio/studio.go:85]: should make sure websocket connecting succeed

**Review by hetao92 — APPROVED** [review]
LGTM

**Pull request merged**